File: Sources/DEXT/MacoswheelsUserClient.swift
#if canImport(DriverKit)

import DriverKit
#if canImport(ConfigPlane)
import ConfigPlane
#endif
#if canImport(WheelProtocol)
import WheelProtocol
#endif
#if canImport(Drivers)
import Drivers
#endif
import os.log

final class MacoswheelsUserClient: IOUserClient {

    weak var driver: MacoswheelsDriver?
    private let log = OSLog(subsystem: "it.allard.macoswheels.dext", category: "UserClient")

    override func externalMethod(_ selector: UInt32,
                                 dispatch: IOUserClientMethodDispatch,
                                 target: AnyObject?,
                                 reference: UnsafeMutableRawPointer?) -> IOReturn
    {
        guard let sel = UserClientSelector(rawValue: selector) else {
            return kIOReturnBadArgument
        }
        guard let driver = driver else { return kIOReturnNotReady }
        switch sel {
        case .getDeviceList:    return handleGetDeviceList(driver: driver)
        case .getInfo:          return handleGetInfo(driver: driver)
        case .getCapabilities:  return handleGetCapabilities(driver: driver)
        case .setRotationRange: return handleSetRange(driver: driver)
        case .setAutocenter:    return handleSetAutocenter(driver: driver)
        case .setGain:          return handleSetGain(driver: driver)
        case .reset:            return handleReset(driver: driver)
        case .vendorCommand:    return kIOReturnUnsupported
        }
    }

    private func handleGetDeviceList(driver: MacoswheelsDriver) -> IOReturn {
        kIOReturnSuccess
    }
    private func handleGetInfo(driver: MacoswheelsDriver) -> IOReturn {
        kIOReturnSuccess
    }
    private func handleGetCapabilities(driver: MacoswheelsDriver) -> IOReturn {
        kIOReturnSuccess
    }
    private func handleSetRange(driver: MacoswheelsDriver) -> IOReturn {
        kIOReturnSuccess
    }
    private func handleSetAutocenter(driver: MacoswheelsDriver) -> IOReturn {
        kIOReturnSuccess
    }
    private func handleSetGain(driver: MacoswheelsDriver) -> IOReturn {
        kIOReturnSuccess
    }
    private func handleReset(driver: MacoswheelsDriver) -> IOReturn {
        do { try driver.session?.driver.initialize(); return kIOReturnSuccess }
        catch { return kIOReturnInternalError }
    }
}

#endif

File: Sources/DEXT/MacoswheelsUserClient.cpp
#include <os/log.h>
#include <stdint.h>
#include <string.h>

#include <DriverKit/IOLib.h>
#include <DriverKit/IOService.h>
#include <DriverKit/IOUserClient.h>
#include <DriverKit/OSNumber.h>

#include "MacoswheelsUserClient.h"
#include "MacoswheelsDriver.h"

#define Log(fmt, ...) \
    os_log(OS_LOG_DEFAULT, "MacoswheelsUserClient: " fmt, ##__VA_ARGS__)

enum {
    kSelectorGetDeviceList    = 0,
    kSelectorGetInfo          = 1,
    kSelectorGetCapabilities  = 2,
    kSelectorSetRotationRange = 3,
    kSelectorSetAutocenter    = 4,
    kSelectorSetGain          = 5,
    kSelectorReset            = 6,
    kSelectorVendorCommand    = 7,
};

struct MacoswheelsUserClient_IVars {
    MacoswheelsDriver *driver;
};

bool MacoswheelsUserClient::init() {
    bool ok = super::init();
    if (!ok) return false;
    ivars = IONewZero(MacoswheelsUserClient_IVars, 1);
    return ivars != NULL;
}

void MacoswheelsUserClient::free() {
    IOSafeDeleteNULL(ivars, MacoswheelsUserClient_IVars, 1);
    super::free();
}

kern_return_t IMPL(MacoswheelsUserClient, Start) {
    kern_return_t ret = Start(provider, SUPERDISPATCH);
    if (ret != kIOReturnSuccess) return ret;
    ivars->driver = OSDynamicCast(MacoswheelsDriver, provider);
    if (ivars->driver == NULL) {
        Log("provider is not MacoswheelsDriver");
        Stop(provider, SUPERDISPATCH);
        return kIOReturnUnsupported;
    }
    return kIOReturnSuccess;
}

kern_return_t IMPL(MacoswheelsUserClient, Stop) {
    ivars->driver = NULL;
    return Stop(provider, SUPERDISPATCH);
}

kern_return_t MacoswheelsUserClient::ExternalMethod(
    uint64_t selector,
    IOUserClientMethodArguments *arguments,
    const IOUserClientMethodDispatch *dispatch,
    OSObject *target,
    void *reference)
{
    MacoswheelsDriver *driver = ivars->driver;
    if (driver == NULL) return kIOReturnNotReady;

    switch (selector) {
    case kSelectorSetRotationRange: {
        if (arguments->scalarInputCount < 1) return kIOReturnBadArgument;
        uint16_t degrees = (uint16_t) arguments->scalarInput[0];
        return driver->SetRotationRange(degrees);
    }
    case kSelectorSetAutocenter: {
        if (arguments->scalarInputCount < 1) return kIOReturnBadArgument;
        uint8_t pct = (uint8_t) arguments->scalarInput[0];
        return driver->SetAutocenter(pct);
    }
    case kSelectorSetGain: {
        if (arguments->scalarInputCount < 1) return kIOReturnBadArgument;
        uint8_t pct = (uint8_t) arguments->scalarInput[0];
        return driver->SetGain(pct);
    }
    case kSelectorReset:
        return driver->ResetWheel();

    case kSelectorGetDeviceList:
    case kSelectorGetInfo:
    case kSelectorGetCapabilities:
    case kSelectorVendorCommand:
        return kIOReturnUnsupported;

    default:
        Log("unknown selector %llu", selector);
        return kIOReturnBadArgument;
    }
}
